#📌 Extracción

In [278]:
import pandas as pd

In [279]:
datos = pd.read_json('/content/TelecomX_Data.json')

In [280]:
datos.info()
datos.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

##Normalizando las columnas anidadas

In [281]:
#df_customer_normalize = pd.json_normalize(datos['customer'])
#df_phone_normalize = pd.json_normalize(datos['phone'])
#df_internet_normalize = pd.json_normalize(datos['internet'])
#df_account_normalize = pd.json_normalize(datos['account'])

In [282]:
datos1 = [pd.json_normalize(datos[col]) for col in ['customer', 'phone', 'internet', 'account']]

##Creando el DataFrame con todas las columnas

In [283]:
#df_normalized = pd.concat([datos[['customerID']],datos[['Churn']],df_customer_normalize, df_phone_normalize, df_internet_normalize, df_account_normalize], axis=1)
#df_normalized.head()

In [284]:
df_final = pd.concat([datos[['customerID']],datos[['Churn']],*datos1], axis=1)
df_final.head()


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [285]:
list(df_final.columns)


['customerID',
 'Churn',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Charges.Monthly',
 'Charges.Total']

##Opcional: guardar el archivo json normalizado

In [286]:
#df_final.to_json('telecomX_normalizado_original.json')

##Diccionario de datos

*    customerID: número de identificación único de cada cliente
*    Churn: si el cliente dejó o no la empresa
*    gender: género (masculino y femenino)
*    SeniorCitizen: información sobre si un cliente tiene o no una edad    igual o mayor a 65 años
*    Partner: si el cliente tiene o no una pareja
*    Dependents: si el cliente tiene o no dependientes
*    tenure: meses de contrato del cliente
*    PhoneService: suscripción al servicio telefónico
*    MultipleLines: suscripción a más de una línea telefónica
*   InternetService: suscripción a un proveedor de internet
*    OnlineSecurity: suscripción adicional de seguridad en línea
*    OnlineBackup: suscripción adicional de respaldo en línea
*    DeviceProtection: suscripción adicional de protección del dispositivo
*    TechSupport: suscripción adicional de soporte técnico, menor tiempo de espera
*    StreamingTV: suscripción de televisión por cable
*    StreamingMovies: suscripción de streaming de películas
*    Contract: tipo de contrato
*    PaperlessBilling: si el cliente prefiere recibir la factura en línea
*    PaymentMethod: forma de pago
*    Charges.Monthly: total de todos los servicios del cliente por mes
*    Charges.Total: total gastado por el cliente



#🔧 Transformación

##Columnas en letra minuscula

In [287]:
df_final.columns = df_final.columns.str.lower()

##Columnas sin espacios

In [288]:
df_final.columns = df_final.columns.str.strip()

In [289]:
df_final.columns

Index(['customerid', 'churn', 'gender', 'seniorcitizen', 'partner',
       'dependents', 'tenure', 'phoneservice', 'multiplelines',
       'internetservice', 'onlinesecurity', 'onlinebackup', 'deviceprotection',
       'techsupport', 'streamingtv', 'streamingmovies', 'contract',
       'paperlessbilling', 'paymentmethod', 'charges.monthly',
       'charges.total'],
      dtype='object')

##Renombrando las columnas

In [290]:
df_final = df_final.rename(columns={'seniorcitizen': 'senior_citizen', 'phoneservice': 'phone_service',
                                    'multiplelines': 'multiple_lines', 'internetservice': 'internet_service',
                                    'onlinesecurity': 'online_security', 'onlinebackup': 'online_backup',
                                    'deviceprotection': 'device_protection', 'techsupport': 'tech_support',
                                    'streamingtv': 'streaming_tv', 'streamingmovies': 'streaming_movies',
                                    'paperlessbilling': 'paperless_billing', 'paymentmethod': 'payment_method',
                                    'charges.monthly': 'monthly_charges', 'charges.total': 'total_charges'})



In [291]:
df_final.columns

Index(['customerid', 'churn', 'gender', 'senior_citizen', 'partner',
       'dependents', 'tenure', 'phone_service', 'multiple_lines',
       'internet_service', 'online_security', 'online_backup',
       'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies',
       'contract', 'paperless_billing', 'payment_method', 'monthly_charges',
       'total_charges'],
      dtype='object')

##Valores unicos de las columnas

In [292]:
columnas = list(df_final.columns)

valores_unicos = {}

for columna in columnas:
    print(f'{columna}: {df_final[columna].unique()}')


customerid: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
churn: ['No' 'Yes' '']
gender: ['Female' 'Male']
senior_citizen: [0 1]
partner: ['Yes' 'No']
dependents: ['Yes' 'No']
tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0]
phone_service: ['Yes' 'No']
multiple_lines: ['No' 'Yes' 'No phone service']
internet_service: ['DSL' 'Fiber optic' 'No']
online_security: ['No' 'Yes' 'No internet service']
online_backup: ['Yes' 'No' 'No internet service']
device_protection: ['No' 'Yes' 'No internet service']
tech_support: ['Yes' 'No' 'No internet service']
streaming_tv: ['Yes' 'No' 'No internet service']
streaming_movies: ['No' 'Yes' 'No internet service']
contract: ['One year' 'Month-to-month' 'Two year']
paperless_billing: ['Yes' 'No']
payment_method: ['Mailed check' 'Electroni

##Determinando Valores vacios en la columna churn

In [293]:
df_final['churn'].value_counts()

churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

In [294]:
vacios = (df_final['churn'] == '').sum()
vacios


np.int64(224)

##Porcentaje de los valores vacios

In [295]:
total_filas_churn = len(df_final)
porcentaje = (df_final['churn'].value_counts() / total_filas_churn) * 100
print(f'{porcentaje.iloc[0]:.2f}% Porcentaje de No')
print(f'{porcentaje.iloc[1]:.2f}% Porcentaje de Yes')
print(f'{porcentaje.iloc[2]:.2f}% Porcentaje de valores vacios ')

71.20% Porcentaje de No
25.72% Porcentaje de Yes
3.08% Porcentaje de valores vacios 


##Vemos que la columna 'churn' contiene 224 datos vacios
**Por que eliminar**:
* Esto representa el 3.08% del total de datos de la columna.
* Es un porcentaje bajo < 5%, no afectara significativamente el analisis.
* Evitamos sesgos en modelos predictivos.

###Elimina los valores vacios '' en la columna churn
* usando indexacion boolena.

In [296]:
# Devuelve el DataFrame donde churn no contiene '' vacios
df_clean = df_final[df_final['churn'] != ''].copy()
df_clean['churn'].value_counts()

churn
No     5174
Yes    1869
Name: count, dtype: int64

##Convertir la columnas total_charges a tipo float

In [297]:
df_clean['total_charges'] = pd.to_numeric(df_clean['total_charges'], errors='coerce')

##Convertir columnas especificas de (Yes) = 1 y (No) = 0

In [298]:
for col in ['churn', 'partner', 'dependents', 'phone_service', 'paperless_billing']:
  df_clean[col] = df_clean[col].map({'Yes': 1, 'No': 0}).astype(int)


##Valores vacios en las otras columnas

In [299]:
for col in columnas:
   print(f'{col}: {df_clean[col].isna().sum()}')

customerid: 0
churn: 0
gender: 0
senior_citizen: 0
partner: 0
dependents: 0
tenure: 0
phone_service: 0
multiple_lines: 0
internet_service: 0
online_security: 0
online_backup: 0
device_protection: 0
tech_support: 0
streaming_tv: 0
streaming_movies: 0
contract: 0
paperless_billing: 0
payment_method: 0
monthly_charges: 0
total_charges: 11


##Identificando la columna = total_charges, tiene datos vacios

In [300]:
faltantes = df_clean[df_clean['total_charges'].isna()]
print(f'Filas faltantes {len(faltantes)}')

Filas faltantes 11


In [301]:
df_filtrado = df_clean[df_clean['total_charges'].isna()]
df_filtrado

customerid  churn  gender  senior_citizen  partner  dependents  tenure  \
975   1371-DWPAZ      0  Female               0        1           1       0   
1775  2520-SGTTA      0  Female               0        1           1       0   
1955  2775-SEFEE      0    Male               0        0           1       0   
2075  2923-ARZLG      0    Male               0        1           1       0   
2232  3115-CZMZD      0    Male               0        0           1       0   
2308  3213-VVOLG      0    Male               0        1           1       0   
2930  4075-WKNIU      0  Female               0        1           1       0   
3134  4367-NUYAO      0    Male               0        1           1       0   
3203  4472-LVYGI      0  Female               0        1           1       0   
4169  5709-LVOEQ      0  Female               0        1           1       0   
5599  7644-OMVMY      0    Male               0        1           1       0   

      phone_service    multiple_lines internet_service  ...  \
975               0  No phone service              DSL  ...   
1775              1                No               No  ...   
1955              1               Yes              DSL  ...   
2075              1                No               No  ...   
2232              1                No               No  ...   
2308              1               Yes               No  ...   
2930              1               Yes              DSL  ...   
3134              1               Yes               No  ...   
3203              0  No phone service              DSL  ...   
4169              1                No              DSL  ...   
5599              1                No               No  ...   

            online_backup    device_protection         tech_support  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                  Yes                   No                  Yes   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                   No                  Yes                  Yes   
4169                  Yes                  Yes                   No   
5599  No internet service  No internet service  No internet service   

             streaming_tv     streaming_movies  contract paperless_billing  \
975                   Yes                   No  Two year                 0   
1775  No internet service  No internet service  Two year                 0   
1955                   No                   No  Two year                 1   
2075  No internet service  No internet service  One year                 1   
2232  No internet service  No internet service  Two year                 0   
2308  No internet service  No internet service  Two year                 0   
2930                  Yes                   No  Two year                 0   
3134  No internet service  No internet service  Two year                 0   
3203                  Yes                   No  Two year                 1   
4169                  Yes                  Yes  Two year                 0   
5599  No internet service  No internet service  Two year                 0   

                 payment_method monthly_charges  total_charges  
975     Credit card (automatic)           56.05            NaN  
1775               Mailed check           20.00            NaN  
1955  Bank transfer (automatic)           61.90            NaN  
2075               Mailed check           19.70            NaN  
2232               Mailed check           20.25            NaN  
2308               Mailed check           25.35            NaN  
2930               Mailed check           7

##Analisis para determinar:

* si los datos faltantes en la columna 'total_charges', tienen relacion con la columuna 'churn' (si el cliente dejó o no la empresa)

In [302]:
# Análisis cruzado: tenure vs churn vs total_charges faltante
if 'tenure' in df_clean.columns:

    # Crear tabla cruzada
    tabla_cruzada = pd.crosstab(
        index=faltantes['tenure'],
        columns=faltantes['churn'],
        margins=True
    )

    tabla_cruzada

    # Clientes con tenure = 0 / meses de contrato
    tenure_0 = faltantes[faltantes['tenure'] == 0]
    if len(tenure_0) > 0:
        print(f"\nMeses de contrato del cliente (tenure) = 0 y total_charges faltantes: {len(tenure_0)}")
        print(f"Si el cliente dejó o no la empresa(Churn), distribución: {tenure_0['churn'].value_counts().to_dict()}")


Meses de contrato del cliente (tenure) = 0 y total_charges faltantes: 11
Si el cliente dejó o no la empresa(Churn), distribución: {0: 11}


##Interpretacion:
* tenure = 0 : Son clientes completamente nuevos (0 meses de antiguedad).
* churn = 0 : Ninguno ha abandonado (todos estan activos).
* total_charges = 11 : Tiene cargos totales faltantes, por que aun no han completado un mes de servicio.
* Estos clientes no han generado cargos aun

In [303]:
#Rellena los valores nulos en la columna total_charges con 0
df_clean['total_charges'] = df_clean['total_charges'].fillna(0)

##Reseteando el indice

In [304]:
df_clean = df_clean.reset_index()

##Opcional: guardar el Dataframe limpio para el analisis

In [305]:
#df_clean.to_json('telecomX_limpio.json')

#📊 Carga y análisis

#📄Informe final